File: LevelSDKExampleApp/app/src/main/java/com/theshopatvsp/levelandroidsdk/ble/model/strategy/PacketParserManager.java
package com.theshopatvsp.levelandroidsdk.ble.model.strategy;

import  com.theshopatvsp.levelandroidsdk.ble.model.constants.DeviceCommand;
import  com.theshopatvsp.levelandroidsdk.ble.model.exception.DataLengthException;
import  com.theshopatvsp.levelandroidsdk.ble.model.exception.SequenceIdException;
import  com.theshopatvsp.levelandroidsdk.ble.model.response.DataPacket;
import  com.theshopatvsp.levelandroidsdk.ble.model.response.RecordData;

/**
 * Created by andrco on 10/2/15.
 */
public class PacketParserManager {
    private static RecordData record;

    public static DataPacket parse(int expectedPacketIdIn, byte packet[]) throws Exception {

        DeviceCommand command = getCommand(expectedPacketIdIn, packet);

        if (command == null) {
            return null;
        }

        if (command == DeviceCommand.START_OF_RECORD) {
            record = new StartOfRecord().parse(packet);

            if( record.isFinished() )
                return record;
        } else if (command == DeviceCommand.RECORD_CONTINUE) {
            record = new StartOfRecord().continueRecord(record, packet);

            if (record.isFinished()) {
                return record;
            }
        } else {
            PacketParser parser = command.getParser();

            if (parser != null) {
                DataPacket data = parser.parse(packet);

                return data;
            }
        }

        return null;
    }

    private static DeviceCommand getCommand(int expectedPacketIdIn, byte packet[]) throws DataLengthException, SequenceIdException {
        int packetIdIn = 0x00 << 24 | packet[0] & 0xff;

        if( packet == null || packet.length < 2 ) {
            throw new DataLengthException("Packet null of length is less then 2.");
        }

        if( expectedPacketIdIn != packetIdIn) {
            throw new SequenceIdException("Packet ID: " + packetIdIn + " does not match " + expectedPacketIdIn);
        }

        return DeviceCommand.getByCommand(packet[1]);
    }
}

File: LevelIOSSdkExample/LevelIOSSdkExample/LevelIOSSdk/model/strategy/PacketParserManager.swift
//
//  PacketParserManager.swift
//  LevelIOSSDK
//
//  Created by Andrew Cook on 6/24/16.
//  Copyright © 2016 TheShop. All rights reserved.
//

import Foundation

enum PacketErrors: ErrorType {
    case DataLength
    case SequenceId
}

class PacketParserManager {
    var factory: DataPacketFactory
    var record: RecordData?
    var setTime: Bool = false
    var timeDiff: Double = 0
    
    init() {
        record = nil
        factory = DataPacketFactory()
    }
    
    func parse(expectedPacketIdIn: Int, packet: [UInt8]) throws -> DataPacket? {
        if packet.count < 2 {
            throw PacketErrors.DataLength
        }
        
        if expectedPacketIdIn != Int(packet[0]) {
            throw PacketErrors.SequenceId
        }
        
        if let command: DeviceCommand = DeviceCommand(rawValue: Int(packet[1])) {
            if command == DeviceCommand.StartOfRecord {
                record = StartOfRecordParser().parse(packet) as? RecordData
                
                if setTime {
                    record!.timestamp = record!.timestamp + timeDiff
                }
                
                if record!.isFinished() {
                    return record
                }
            } else if command == DeviceCommand.RecordContinue {
                record!.continueRecord(packet)
                
                if record!.isFinished() {
                    return record
                }
            } else {
                if let parser: DataPacketParserProtocol = factory.getDataPacket(packet) {
                    let packet: DataPacket = parser.parse(packet)!
                    
                    if packet is TimePacket {
                        let timePacket: TimePacket = packet as! TimePacket
                        
                        if abs(timePacket.timestamp - timePacket.received.timeIntervalSince1970) > 2 * 60 * 1000 {
                            setTime = true
                            timeDiff = timePacket.received.timeIntervalSince1970 - timePacket.timestamp
                        }
                    }
                    
                    return packet
                }
            }
        }
        
        return nil
    }
}
